Framework/Core/src/DataProcessingDevice.cxx

```diff
@@ -96,9 +91,26 @@ void DataProcessingDevice::Reset() { mServiceRegistry.get<CallbackService>()(Cal
 /// is divided in two parts. In the first one we define a set of lambdas
 /// which describe what is actually going to happen, hiding all the state
 /// boilerplate which the user does not need to care about at top level.
-/// In the second part 
-bool
-DataProcessingDevice::HandleData(FairMQParts &iParts, int /*index*/) {
+/// In the second part
+bool DataProcessingDevice::ConditionalRun()
+{
+  static const auto reftime = std::chrono::system_clock::now();
+
+  FairMQParts parts;
+
+  for (auto& channel : mInputChannels) {
+    auto result = this->ReceiveAsync(parts, channel.name);
+    if (result > 0) {
```

Comment:
wouldn't that mean that, if the input queue of one channel is filled, subsequent channels are never read?

Reply:
Correct. This is (wrongly) assuming that the first channel will not fill the pipes. Will fix it.